languages/java/app-encryption/src/test/java/com/godaddy/asherah/appencryption/SessionFactoryTest.java

```diff
@@ -42,32 +49,677 @@
+      sessionCache = sessionFactory.getSessionCache();
+
+      try (Session<byte[], byte[]> session = sessionFactory.getSessionBytes("1234")) {
+        assertNotNull(session);
+      }
+      // Verify nothing evicted yet
+      assertTrue(sessionCache.estimatedSize() > 0);
+    }
+    // Verify closing the factory invalidated and cleaned up entries
+    assertTrue(sessionCache.estimatedSize() == 0);
+  }
+
+  @SuppressWarnings("unchecked")
+  @Test
+  void testSessionCacheGetSessionWhileStillUsedAndNotExpiredShouldNotEvict() {
+    CryptoPolicy policy = BasicExpiringCryptoPolicy.newBuilder()
+        .withKeyExpirationDays(1)
+        .withRevokeCheckMinutes(30)
+        .withCanCacheSessions(true)
+        .build();
+
+    try (SessionFactory factory = SessionFactory.newBuilder(testProductId, testServiceId)
+        .withMetastore(metastoreSpy)
+        .withCryptoPolicy(policy)
+        .withStaticKeyManagementService(testMasterKey)
+        .build()) {
+      try (Session<byte[], byte[]> session = factory.getSessionBytes(testPartitionId)) {
+        byte[] payload = new byte[] {0, 1, 2, 3, 4, 5, 6, 7};
+        byte[] drr = session.encrypt(payload);
+
+        // Reset so we can examine 2nd session's interactions
+        reset(metastoreSpy);
+
+        // Use same partition to get the same cached session while it's still in use
+        try (Session<byte[], byte[]> session2 = factory.getSessionBytes(testPartitionId)) {
+          byte[] decryptedPayload = session2.decrypt(drr);
+
+          assertArrayEquals(payload, decryptedPayload);
+
+          // verify no metastore interactions in the decrypt flow (since IKs cached via session caching)
+          verify(metastoreSpy, never()).load(any(), any());
+        }
+      }
+    }
+  }
+
+  @SuppressWarnings("unchecked")
+  @Test
+  void testSessionCacheGetSessionWhileStillUsedAndExpiredShouldNotEvict() {
+    long sessionCacheExpireMillis = 10;
+    CryptoPolicy policy = spy(BasicExpiringCryptoPolicy.newBuilder()
+        .withKeyExpirationDays(1)
+        .withRevokeCheckMinutes(30)
+        .withCanCacheSessions(true)
+        .build());
+    when(policy.getSessionCacheExpireMillis()).thenReturn(sessionCacheExpireMillis);
+
+    try (SessionFactory factory = SessionFactory.newBuilder(testProductId, testServiceId)
+        .withMetastore(metastoreSpy)
+        .withCryptoPolicy(policy)
+        .withStaticKeyManagementService(testMasterKey)
+        .build()) {
+      try (Session<byte[], byte[]> session = factory.getSessionBytes(testPartitionId)) {
+        byte[] payload = new byte[] {0, 1, 2, 3, 4, 5, 6, 7};
+        byte[] drr = session.encrypt(payload);
+
+        // Sleep to trigger the expiration
+        // TODO Consider refactoring the Caffeine usage to allow injecting a ticker if sleeping becomes unreliable
+        try {
+          Thread.sleep(sessionCacheExpireMillis * 3);
+        } catch (Exception e) {
+          fail(e.getMessage());
+        }
+
```

Comment:
can verify here that even after timeout, we have 1 entry in the cache

Reply:
that's what the metastore interaction verify is effectively testing, but i can add the extra check